src/cli/lint/internalLints/refreshConfigFiles.ts

```diff
@@ -5,44 +5,83 @@ import { writeFile } from 'fs-extra';
 import stripAnsi from 'strip-ansi';
 
 import type { Logger } from '../../../utils/logging';
-import { detectPackageManager } from '../../../utils/packageManager';
+import {
+  type PackageManagerConfig,
+  detectPackageManager,
+} from '../../../utils/packageManager';
 import { readBaseTemplateFile } from '../../../utils/template';
 import { getDestinationManifest } from '../../configure/analysis/package';
 import { createDestinationFileReader } from '../../configure/analysis/project';
 import { mergeWithIgnoreFile } from '../../configure/processing/ignoreFile';
 import type { InternalLintResult } from '../internal';
 
-const REFRESHABLE_IGNORE_FILES = [
-  '.eslintignore',
-  '.gitignore',
-  '.prettierignore',
+const ensureNoNpmrcExclusion = (s: string) =>
+  s
+    .split('\n')
+    .filter((line) => line.includes('!') || !line.includes('.npmrc'))
+    .join('\n');
+
+const ensureNoAuthToken = (s: string) =>
+  s
+    .split('\n')
+    .filter((line) => !line.includes('authToken'))
+    .join('\n');
+
+const REFRESHABLE_CONFIG_FILES = [
+  { name: '.eslintignore' },
+  {
+    name: '.gitignore',
+    additionalMapping: (
+      contents: string,
+      packageManager: PackageManagerConfig,
+    ) => {
+      if (packageManager.command === 'pnpm') {
+        return ensureNoNpmrcExclusion(contents);
+      }
+      return contents;
+    },
+  },
```

Comment:
The remaining concern I have is:

1. An existing project like https://github.com/seek-oss/koala/blob/e3278b52be501f891897c5cf2621657369e9b285/.gitignore#L17 uses Yarn and relies on skuba to gitignore `.npmrc`
2. An engineer has copied their `.npmrc` + auth token into the local project directory for testing purposes
3. When this new version applies its config refresh, `mergeWithIgnoreFile` will remove the `.npmrc` gitignore
4. The engineer then commits and pushes their local `.npmrc` + auth token

My initial thought was that this should be _preprocessed_, something like checking if `.npmrc` was previously gitignored and then preserving it at the bottom of the gitignore file, before `mergeWithIgnoreFile` goes and updates the skuba managed section. It could also be implemented as a patch if we can get that to run before the config refresh.

Reply:
Hmm I did a test. Does it fully cover the scenarios?

I created a test repository, and did the following:

- Set up a `.gitignore` file with `mysecretfile`. Pushed it. Created a `mysecretfile` with a secret in it. 
- In a separate checkout, modified .gitignore to remove that entry, and in the same commit, created a mysecretfile with fake "managed" content. Pushed.
- Pulled in the original checkout, and the new managed content overwrote the file and my secret was deleted.

In this scenario, it would be quite hard to accidentally push the secret. 

The cases where I can see it happening: 

- Not all files are committed (I could see this potentially happening if doing `git commit -am` not picking up new files). I suppose leaving it up to manual intervention helps this somewhat, eg. by adding a "# TODO: remove this line and make sure you push the npmrc"?
- If for some reason we don't write npmrc and gitignore (e.g. one crashes). This seems unlikely. 

I _am_ open to workshopping this, just want to understand what we're trying to guard against. What are your thoughts?

Reply:
All good, I may be mistaken but I think the scenario is that

- Bullet two runs in CI so the engineer isn't necessarily looking at what's going on
- Bullet two doesn't synthesise a conflicting `mysecretfile` because the project is still Yarn based

Reply:
Ah, I see, of course 🤦 

A way to tackle that I suppose is to just always have the npmrc synthesised? I.e. reverting e951645a72d8b82d1146b05fb38da4c2bdea6a5a - I can't [really see it causing too many issues](https://github.com/search?q=%28org%3Anpm+OR+org%3Ayarnpkg%29+public-hoist-pattern&type=code) with other package managers, but hard to entirely prove